LoginClient/PipeClient.cs

```diff
@@ -0,0 +1,368 @@
+
+        public bool IsLoggedIn()
+        {
+            return isLoggedIn;
+        }
+
+        private void IsLoggedIn(bool isLoggedInOrNot)
+        {
+            isLoggedIn = isLoggedInOrNot;
+
+            if (isLoggedIn)
+            {
+                ConnectedToServer();
+            }
+        }
+
+        /// <summary>
+        /// Returns true if client is connected to a server.
+        /// </summary>
+        /// <returns>Boolean</returns>
+        public bool IsConnected()
+        {
+            return connected;
+        }
+
+        /// <summary>
+        /// Set the value of the connected boolean.
+        /// </summary>
+        /// <param name="b">The new value of connected.</param>
+        private void IsConnected(bool b)
+        {
+            connected = b;
+        }
+
+        /// <summary>
+        /// Return the name of the pipe used to connect to the server.
+        /// </summary>
+        /// <returns>string</returns>
+        public string GetPipeName()
+        {
+            return pipeName;
+        }
+
+        /// <summary>
+        /// Set the name of the pipe that connects to the server.
+        /// </summary>
+        /// <param name="newPipeName">The new pipe name.</param>
+        private void SetPipeName(string newPipeName)
+        {
+            pipeName = newPipeName;
+        }
+
+        #endregion
+
+
+        #region Connection
+
+        /// <summary>
+        /// Disconnects the client from the server.
+        /// </summary>
+        public void Disconnect()
+        {
+            if(!IsConnected())
+            {
+                return;
+            }
+
+            // We're no longer connected to the server.
+            IsConnected(false);
+            SetPipeName(null);
+            IsLoggedIn(false);
+
+            readThread.Abort();
+
+            // Clean up the resources
+            if (stream != null)
+            {
+                stream.Close();
+            }
+            handle.Close();
+
+            stream = null;
+            handle = null;
+        }
+
+        /// <summary>
+        /// Connects client to the server with the pipename.
+        /// </summary>
+        /// <param name="pipename">The name of the pipe to connect to.</param>
+        public void Connect(string pipename)
+        {
+            if (IsConnected())
+            {
+                throw new Exception("Already connected to server.");
+            }
+
+            SetPipeName(pipename);
+
+            handle = CreateFile(
+                GetPipeName(),
+                0xC0000000, // GENERIC_READ | GENERIC_WRITE = 0x80000000 | 0x40000000
+                0,
+                IntPtr.Zero,
+                3,  // OPEN_EXISTING
+                0x40000000, // FILE_FLAG_OVERLAPPED
+                IntPtr.Zero);
+
+            // Couldn't create a handle.
+            // The server is probably not running.
+            if (handle.IsInvalid)
+            {
+                MessageBox.Show(
```

Comment:
Not a Bug.
Code recommendation. Don't fix until bug found.
Non form classes, generally, shouldn;t use form stuff like this.
Throw an exception, catch it in the form, show the message there.